packages/sdk/schema/src/types/view.ts

```diff
@@ -2,33 +2,83 @@
 // Copyright 2024 DXOS.org
 //
 
-import { type MutableSchema } from '@dxos/echo-schema';
-import { getProperty } from '@dxos/effect';
-import { invariant } from '@dxos/invariant';
-
-import { schemaForType } from './field';
-import { type FieldType, type ViewType } from './types';
-
-export const addFieldToView = (
-  schema: MutableSchema,
-  view: ViewType,
-  field: FieldType,
-  position = view.fields.length,
-) => {
-  invariant(!getProperty(schema, field.path), `Field path already exists in schema: ${field.path}`);
-  invariant(position >= 0 && position <= view.fields.length, `Invalid field position: ${position}`);
-
-  const propertySchema = schemaForType[field.type];
-  invariant(propertySchema, `No schema for field type: ${field.type}`);
-
-  schema.addFields({ [field.path]: propertySchema });
-  view.fields.splice(position, 0, field);
-};
-
-export const removeFieldFromView = (schema: MutableSchema, view: ViewType, field: FieldType) => {
-  const index = view.fields.findIndex((f) => f.path === field.path);
-  if (index !== -1) {
-    view.fields.splice(index, 1);
-  }
-  schema.removeFields([field.path]);
-};
+import { JsonSchemaType, QueryType } from '@dxos/echo-schema';
+import { S } from '@dxos/effect';
+
+import { FieldKindEnum } from './annotations';
+
+// TODO(burdon): Pattern for error IDs (i.e., don't put user-facing messages in the annotation).
+export const PathSchema = S.String.pipe(
+  S.nonEmptyString({ message: () => 'Property is required.' }),
+  S.pattern(/^[a-zA-Z_$][\w$]*(?:\.[a-zA-Z_$][\w$]*)*$/, { message: () => 'Invalid property path.' }),
+);
+
+/**
+ * Stored field metadata.
+ */
+export const FieldSchema = S.mutable(
+  S.Struct({
+    path: PathSchema,
```

Comment:
seems extraneos to have path both on `field` and `schema` parts of the view. perhaps fields in the array should just reference schema fields by id?